src/Npgsql/NpgsqlDataReader.cs

```diff
@@ -884,10 +884,17 @@ internal async Task Cleanup(bool async, bool connectionClosing=false)
             // on .NET Framework.
             if (_sendTask != null)
             {
-                if (async)
-                    await _sendTask;
-                else
-                    _sendTask.GetAwaiter().GetResult();
+                try
+                {
+                    if (async)
+                        await _sendTask;
+                    else
+                        _sendTask.GetAwaiter().GetResult();
+                }
+                catch (OperationCanceledException)
+                {
+                    // We've cancelled the task, because the cancellation request from the NpgsqlCommand was not successful - ignoring the exception
```

Comment:
I probably need a bit more of elaboration to really understand this.

Is this the situation where we tried to cancel a task but came too late and it had already finished all potentially blocking calls (`FlushAsync()`) so we are no longer interested in the `OperationCanceledException` and move on?

If so, how do we know that this happened and that it isn't just a normal successful cancellation where the user expects an `OperationCanceledException`?

Reply:
> Is this the situation where we tried to cancel a task but came too late and it had already finished all potentially blocking calls (`FlushAsync()`) so we are no longer interested in the `OperationCanceledException` and move on?

More like the opposite. User requested the cancellation, we've send the cancellation request. If it was successful, it should work as it is. But if it was not, the task should be cancelled through the cancellation token (this will be done in next pr).

Reply:
Essentially, there would be another `CancellationTokenSource`, from which we pass the token to the reader and the writer. And it should be cancelled only if the cancellation request has failed.

Reply:
Just to be on a safe side, we may check in the write task, if the cancellation token was really cancelled. And if it was not, throw a NpgsqlException with the original exception as inner exception.